pytorch_lightning/callbacks/model_checkpoint.py

```diff
@@ -239,6 +243,21 @@ def on_validation_end(self, trainer, pl_module) -> None:
             return
         self.save_checkpoint(trainer)
 
+    def on_train_end(self, trainer, *args, **kwargs) -> None:
+        """
+        checkpoints can be saved at the end of the trianing
+        """
+        if not self._trigger_on_train_end:
+            return
+        # as we advance one step at end of training, we use global_step - 1
+        # to avoid saving duplicates
+        trainer.global_step -= 1
+        if (not self._should_skip_saving_checkpoint(trainer) and trainer.checkpoint_connector.has_trained):
+            if self.save_last and self.verbose:
+                rank_zero_info("Saving last checkpoint...")
+            self.save_checkpoint(trainer, is_on_train_end=True)
```

Comment:
@ananthsub ,

> can this not directly call self._save_last_checkpoint ? I think the most common case will be for saving a last.ckpt file at the end of training. this way we don't thread through the is_on_train_end flag everywhere

we could directly call `self._save_last_checkpoint` by ignoring the topK setup.

One thing to discuss is if `trigger_on_train_end` is set, should we guarantee to save `last.ckpt` even if `save_last` is not set?

Reply:
I think we should respect what's set on the callback. The other reason is if we have multiple checkpoint callbacks, we don't need them all to save on train end. We'll configure only one of them to have `save_last=True`